src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/utils/CardinalityEstimation.java

```diff
@@ -0,0 +1,291 @@
+        }
+
+        final Integer cachedCard = cardinalitiesCache.get(pp);
+        if ( cachedCard != null ) {
+            return cachedCard;
+        }
+
+        final PhysicalPlan pp1 = pp.getSubPlan(0);
+        final PhysicalPlan reqBGP = formRequestBasedOnBGPofBGPAdd( (LogicalOpBGPAdd) lop );
+
+        final int cardinality = joinCardinality( pp1, reqBGP );
+        cardinalitiesCache.add( pp, cardinality );
+
+        return cardinality;
+    }
+
+    protected int joinCardinality( final PhysicalPlan pp1, final PhysicalPlan pp2 ) throws FederationAccessException, QueryOptimizationException {
+        final Set<Var> certainJoinVars = ExpectedVariablesUtils.intersectionOfCertainVariables( pp1.getExpectedVariables(), pp2.getExpectedVariables() );
+        final Set<Var> possibleJoinVars = ExpectedVariablesUtils.unionOfAllVariables( pp1.getExpectedVariables(), pp2.getExpectedVariables() );
+        possibleJoinVars.removeAll(certainJoinVars);
+        final Set<Var> allCertainVars = ExpectedVariablesUtils.unionOfCertainVariables( pp1.getExpectedVariables(), pp2.getExpectedVariables() );
+
+        int cardinality = 0;
+
+        if ( !certainJoinVars.isEmpty() ){
+            for ( final Var v : certainJoinVars ) {
+                int c1 = getCardinalityEstimationOfSpecificVar( pp1, v );
+                int c2 = getCardinalityEstimationOfSpecificVar( pp2, v );
+                int c = min(c1, c2);
+                cardinality = max(c, cardinality);
+            }
+        }
+        else if ( !possibleJoinVars.isEmpty() ){
+            for ( final Var v : possibleJoinVars ) {
+                int c1 = getCardinalityEstimationOfSpecificVar( pp1, v );
+                int c2 = getCardinalityEstimationOfSpecificVar( pp2, v );
+                int c = min(c1, c2);
+                cardinality = max(c, cardinality);
+            }
+        }
+        else {
+            for ( final Var v : allCertainVars ) {
+                int c ;
+                if ( pp1.getExpectedVariables().getCertainVariables().contains(v) ){
+                    c = getCardinalityEstimationOfSpecificVar( pp1, v );
+                }
+                else c = getCardinalityEstimationOfSpecificVar( pp2, v );
+                cardinality = max(c, cardinality);
+            }
+        }
+
+        return cardinality;
+    }
+
+    public int getCardinalityEstimationOfSpecificVar( final PhysicalPlan pp, final Var v ) throws FederationAccessException, QueryOptimizationException {
+        final Integer varCachedCard = varSpecificCardinalitiesCache.get(pp, v);
+        if ( varCachedCard != null ) {
+            return varCachedCard;
+        }
+
+        final PhysicalOperatorForLogicalOperator pop = (PhysicalOperatorForLogicalOperator) pp.getRootOperator();
+        final LogicalOperator lop = pop.getLogicalOperator();
+
+        int cardinality = 0;
+        if (lop instanceof LogicalOpRequest) {
+            cardinality = getCardinalityEstimationOfLeafNode(pp );
```

Comment:
Is this part really correct? It may very well be! I just observe that this part is completely independent of the given variable v, which _might_ be a mistake, or not. Please double-check.